logsmanagement/parser.c

```diff
@@ -0,0 +1,1622 @@
+                        log_line_parsed->port = WEB_LOG_INVALID_PORT;
+                        log_line_parsed->parsing_errors++;
+                    }
+                }
+            }
+            else{
+                #if ENABLE_PARSE_WEB_LOG_LINE_DEBUG
+                error("Error while extracting PORT from string");
+                #endif
+                log_line_parsed->port = WEB_LOG_INVALID_PORT;
+                log_line_parsed->parsing_errors++;
+            }
+            #if ENABLE_PARSE_WEB_LOG_LINE_DEBUG
+            debug(D_LOGS_MANAG, "Extracted PORT:%d", log_line_parsed->port);
+            #endif
+
+            goto next_item;
+        }
+
+        if(fields_format[i] == REQ_SCHEME){
+            #if ENABLE_PARSE_WEB_LOG_LINE_DEBUG
+            debug(D_LOGS_MANAG, "Item %d (type: REQ_SCHEME):%.*s", i, (int)field_size, field);
+            #endif
+
+            if(unlikely(field[0] ==  '-' && field_size == 1)){
+                log_line_parsed->req_scheme[0] = '\0';
+                log_line_parsed->parsing_errors++;
+                goto next_item;
+            }
+
+            // field_size = field_size > REQ_SCHEME_MAX_LEN ? REQ_SCHEME_MAX_LEN : field_size;
+            snprintf( log_line_parsed->req_scheme, 
+                      field_size > (REQ_SCHEME_MAX_LEN - 1) ? REQ_SCHEME_MAX_LEN : (field_size + 1), 
+                      "%.*s", (int) field_size, field); 
+
+            if(verify){
+                if(unlikely( strcmp(log_line_parsed->req_scheme, "http") && 
+                             strcmp(log_line_parsed->req_scheme, "https"))){
+                    #if ENABLE_PARSE_WEB_LOG_LINE_DEBUG
+                    error("REQ_SCHEME is invalid (must be either 'http' or 'https')");
+                    #endif
+                    log_line_parsed->req_scheme[0] = '\0';
+                    log_line_parsed->parsing_errors++;
+                }
+            }
+            #if ENABLE_PARSE_WEB_LOG_LINE_DEBUG
+            debug(D_LOGS_MANAG, "Extracted REQ_SCHEME:%s", log_line_parsed->req_scheme);
+            #endif
+            goto next_item;
+        }
+
+        if(fields_format[i] == REQ_CLIENT){
+            #if ENABLE_PARSE_WEB_LOG_LINE_DEBUG
+            debug(D_LOGS_MANAG, "Item %d (type: REQ_CLIENT):%.*s", i, (int)field_size, field);
+            #endif
+
+            if(unlikely(field[0] ==  '-' && field_size == 1)){
+                log_line_parsed->req_client[0] = '\0';
+                log_line_parsed->parsing_errors++;
+                goto next_item;
+            }
+
+            // field_size = field_size > REQ_CLIENT_MAX_LEN ? REQ_CLIENT_MAX_LEN : field_size;
+            snprintf(log_line_parsed->req_client, 
+                     field_size > (REQ_CLIENT_MAX_LEN - 1) ? REQ_CLIENT_MAX_LEN : (field_size + 1), 
+                     "%.*s", (int)field_size, field);
+
+            if(verify){
+                int regex_rc = regexec(&req_client_regex, log_line_parsed->req_client, 0, NULL, 0);
```

Comment:
Regex sometimes are so expensive. It is not necessary to bring for this PR that is already big, but did you take a look in our [eval](https://github.com/netdata/netdata/tree/master/libnetdata/eval) library? Probably it is a better way to address the same necessities here.

Reply:
I didn't know of `eval`, I just had a quick look and it seems like there are a few places I could make use of it. But as you say, probably not in this PR, something to deal with in the future (I think I'll still need POSIX regexs in certain places).

Reply:
All right,  I won't resolve this for we do not forget.

Reply:
ping @Dim-P